docs/proposals/13-seed-management.md

```diff
@@ -0,0 +1,91 @@
+* The `gardenlet` seed controller updates the seed status with the capacity of each resource and how much of it is currently available, using `capacity` and `allocatable` fields that are very similar to the corresponding fields in [the Node status](https://github.com/kubernetes/api/blob/2c3c141c931c0ab1ce1396c3152c72852b3d37ee/core/v1/types.go#L4582-L4593).
+* When scheduling shoots, `gardener-scheduler` is influenced by the allocatable vs. capacity score of the seed. In the simplest possible implementation, it never schedules shoots to a seed that has already reached its maximum capacity for a resource requested by the shoot.
+
+Initially, the only resource considered would be the maximum number of shoots that can be scheduled on a seed. Later, more resources could be added, as well as requests for such resources by shoots. For example, if a shoot requests (via a `resources` field) a certain number of persistent volumes, but the seed's `allocatable` value for persistent volumes is be lower, the shoot can't be scheduled on that seed.
+
+In addition, certain provider extensions could also be extended to update the seed status with the capacity and usage of provider-specific resources, for example `azure-load-balancers` on Azure.
+
+Example seed status with `capacity` and `allocatable` fields:
+
+```yaml
+status:
+  capacity:
+    shoots: 100
+    azure-load-balancers: 300
+  allocatable:
+    shoots: 42
+    azure-load-balancers: 13
+```
+
+## Auto-scaling Gardens by Adding / Removing Seeds
+
+A third important automated seed management scenario is making sure that a Garden cluster has enough seeds registered to schedule new shoots and reschedule shoots from drained seeds without exceeding the seeds capacity for shoots, but not more than actually needed at any given moment. This would involve introducing an auto-scaling mechanism for seeds in Garden clusters. 
+
+One way to achieve this would be to introduce a new custom resource named for example `SeedSet` that represents a certain number of seeds scalable from a minimum to a maximum value, similarly to a replica set for pods. The `SeedSet` resource could have a `scale` subresource, which would allow the existing [Horizontal Pod Autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) Kubernetes controller to scale seedsets via the `HorizontalPodAutoscaler` resource, for example as described in [Horizontal Pod Autoscaler Walkthrough](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/). 
+
+**Note:** Even if the controller is called *Horizontal Pod Autoscaler*, it is capable of scaling any resource with a `scale` subresource, using any custom metric. In this case, the custom metric used for scaling should be the number of shoots already scheduled on that seed either as a direct value or as a percentage of the seed's capacity for shoots introduced above (*utilization*). 
```

Comment:
I'm not too deep into these details, but it sounds like a pretty clever idea.
My first impression would be that this whole thing should be outside of g/g core in a separate component. Do you think the same?

Reply:
I added a new paragraph to clarify this in some more detail: 

> The actual scaling (changing the number of seeds based on the value of the `replicas` field) could be performed by a new `SeedSet` controller that could be part of `gardener-controller-manager` or delivered independently from Gardener core. The new seeds created by this controller should be "shooted seeds" to take full advantage of Gardener's capabilities.

I thought `gardener-controller-manager` is probably a good place to host this new controller, why would you rather do it in a separate component?

Reply:
If it's that small like it sounds then maybe yes. I'm just afraid that we put too much in the core. Anyways, I guess this plays together with what is discussed in https://github.com/gardener/gardener/issues/2181#issuecomment-707516641, and such thing would definitely make sense to be in the core. Let's discuss this again when more concrete implementation details exist.

Reply:
I had a quick look into the code of the HPA controller and it indeed might be technically possible to achieve what you proposed.
Nevertheless, I would want to avoid reusing HPA for scaling Seeds for the following reasons:
- k/k might change details of HPA that can affect us negatively/unexpectedly
- we would have to rely only on custom metrics for scaling Seeds, which might not be sufficient. I.e. the clusterautoscaler doesn't only look at the capacity of Nodes, but also if the pending Pods can actually be scheduled when adding new nodes in a specific node pool (e.g. taints and tolerations, node labels and node selectors, resource requests, etc.). I know, we probably won't be adding such logic in the near future, but I don't won't to run into the wrong direction.
- we would need to register a custom metrics server in the garden cluster. If someone uses a plain cluster as the garden cluster and already has a custom metrics server registered, this will also not work.

Reply:
@timebertt What is the alternative that you propose, having a custom HPA controller just for this purpose, or something else? I can mention this in the document, we should discuss both alternatives in a meeting.

Reply:
I think the alternative would be to introduce a dedicated `seed-autoscaler` component just like the `clusterautoscaler`.

Reply:
OK, added a paragraph to mention this alternative:

> As an alternative to using *Horizontal Pod Autoscaler* a new custom controller could be introduced named for example `seed-autoscaler`, with a similar behavior with respect to seeds and `SeedSets` as the existing `cluster-autoscaler` with respect to `Machines` and `MachineSets`.